### DISCUSSION TITLE
How to run a .pyc file?

### DISCUSSION THREAD

**Discussion by Fluffyalien1422** [PyO3/pyo3#3480] [Questions] [answered]
#2360 says to use `PyModule::import` but it looks like that can only be used to import modules (unless there's something I'm missing)

I'm looking for something similar to `PyModule::from_code` but for .pyc files
I tried using `PyModule::from_code` on a .pyc file (with the `include_bytes` macro) but it doesn't work

How do you do it, if it's possible?

**Comment by mejrs** [comment]
> unless there's something I'm missing

Importing a module will execute that module.

**Reply by Fluffyalien1422** [reply]
I know but you can't pass a file to it, only a module name

**Reply by davidhewitt** [reply, marked as answer]
IIRC `.pyc` is a valid module extension, so e.g. `path/to/code.pyc` should work with `PyModule::import("path.to.code")`.

**Reply by Fluffyalien1422** [reply]
@davidhewitt This works, but it would be preferable if I could manually pass in the file contents. Something like this:
```rs
let module = PyModule::from_pyc(py, include_bytes!("my_file.pyc"));
```

**Reply by mejrs** [reply]
If there is a C api function for that I wouldn't mind, but I don't see any.

**Reply by Fluffyalien1422** [reply]
@mejrs 
> If there is a C api function for that I wouldn't mind, but I don't see any.

This code works!
```rs
Python::with_gil(|_py| {
    let data = &include_bytes!("t.pyc")[16..];
    unsafe {
        let c_data_ptr = data.as_ptr() as *const i8;
        let obj = ffi::PyMarshal_ReadObjectFromString(c_data_ptr, data.len() as isize);

        let _mptr = ffi::PyImport_ExecCodeModuleEx(
            "t".as_ptr() as *const i8,
            obj,
            "t.pyc".as_ptr() as *const i8,
        );

        ffi::Py_DECREF(obj);
    }
});
```

I'll try to implement this and make a PR (#3483)

You can see how this works by looking at the CPython repo
At [run_pyc_file in Python/pythonrun.c](https://github.com/python/cpython/blob/main/Python/pythonrun.c#L1742)
Specifically these lines:
```c
    magic = PyMarshal_ReadLongFromFile(fp);
    if (magic != PyImport_GetMagicNumber()) {
        if (!PyErr_Occurred())
            PyErr_SetString(PyExc_RuntimeError,
                       "Bad magic number in .pyc file");
        goto error;
    }
    /* Skip the rest of the header. */
    (void) PyMarshal_ReadLongFromFile(fp);
    (void) PyMarshal_ReadLongFromFile(fp);
    (void) PyMarshal_ReadLongFromFile(fp);
```
